node/src/Pos/GState/BlockExtra.hs

```diff
@@ -153,6 +154,29 @@ loadUpWhile morph start condition = OldestFirst . reverse <$>
         (\l e -> pure (e : l))
         []
 
+-- | Bottom-top hashes traversal, basically iterating forward links
+-- with condition. Simpler/quicker, doesn't deserialize blunds.
+loadHashesUpWhile
+    :: forall m a. (HasHeaderHash a, MonadDBRead m)
+    => a
+    -> (HeaderHash -> Int -> Bool)
+    -> m (NewestFirst [] HeaderHash)
+loadHashesUpWhile start condition =
+    NewestFirst <$> loadUpWhileDo (headerHash start) 0 []
+  where
+    loadUpWhileDo ::
+           HeaderHash
+        -> Int
+        -> [HeaderHash]
+        -> m [HeaderHash]
+    loadUpWhileDo curH height !res = do
+        mbNextLink <- resolveForwardLink curH
+        if | not (condition curH height) -> pure res
+           | Just nextLink <- mbNextLink ->
+               loadUpWhileDo nextLink (succ height) (curH:res)
+           | otherwise -> pure res
```

Comment:
Why not
```
+loadHashesUpWhile :: (MonadBlockDBRead m, HasHeaderHash a)
+    => a
+    -> (HeaderHash -> Int -> Bool)
+    -> m (OldestFirst [] HeaderHash)
+loadHashesUpWhile start condition =
+    loadUpWhile (pure . Just) identity start condition
```
?

Reply:
Because `loadUpWhile` uses `foldlUpWhileM` which uses `blkGetBlund` thus getting extra data we don't need. I realized it only while rebasing on cardano-sl-1.0 in fact hehe.

Reply:
Ah ok. It might be better to add yet another argument to `foldUpWhileM` to avoid doing `blkGetBlund` unconditionally there 😜

Reply:
Yes, agreed. Let's abstract. I like overabstractions.